pkg/controller/seldondeployment/seldondeployment_controller.go

```diff
@@ -699,51 +403,68 @@ func createComponents(r *ReconcileSeldonDeployment, mlDep *machinelearningv1alph
-				}
-				//overwrite annotations with predictor annotations
-				for _, ann := range p.Annotations {
-					deploy.Spec.Template.Annotations[ann] = p.Annotations[ann]
+			// create services for each container
+			for k := 0; k < len(cSpec.Spec.Containers); k++ {
+				var con *corev1.Container
+				con = utils.GetContainerForDeployment(deploy, cSpec.Spec.Containers[k].Name)
+
+				// engine will later get a special predictor service as it is entrypoint for graph
+				// and no need to expose tfserving container as it's accessed via proxy
+				if con.Name != "seldon-container-engine" && con.Name != "tfserving" && con.Name != "" {
+
+					// service for hitting a model directly, not via engine
+					svc := createContainerService(deploy, p, mlDep, con, c)
+					if svc != nil {
+						c.services = append(c.services, svc)
+					} else {
+						// a user-supplied container may not be a pu so we may not create service for that
+						log.Info("Not creating container service for " + con.Name)
+					}
 				}
-				// Add prometheus annotations
-				deploy.Spec.Template.Annotations["prometheus.io/path"] = getEnv("ENGINE_PROMETHEUS_PATH", "/prometheus")
-				deploy.Spec.Template.Annotations["prometheus.io/port"] = strconv.Itoa(engine_http_port)
-				deploy.Spec.Template.Annotations["prometheus.io/scrape"] = "true"
-
-				deploy.ObjectMeta.Labels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-				deploy.Spec.Selector.MatchLabels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-				deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-
 			}
+			if len(deploy.Spec.Template.Spec.Containers) > 0 && deploy.Spec.Template.Spec.Containers[0].Name != "" {
+				// Add deployment, provided we have a non-empty spec
+				c.deployments = append(c.deployments, deploy)
+			}
+		}
 
-			// create services for each container
-			for k := 0; k < len(cSpec.Spec.Containers); k++ {
-				con := cSpec.Spec.Containers[0]
+		err = createStandaloneModelServers(r, mlDep, &p, &c, p.Graph)
+		if err != nil {
+			return nil, err
+		}
 
-				svc := createContainerService(deploy, p, mlDep, con, c)
+		// Add service orchestrator to engine deployment if needed
+		if !hasSeparateEnginePod {
+			var deploy *appsv1.Deployment
+			found := false
 
-				c.services = append(c.services, svc)
+			pu := machinelearningv1alpha2.GetEnginePredictiveUnit(p.Graph)
+			if pu == nil {
+				return nil, fmt.Errorf("Engine not separate and no pu with localhost service - not clear where to inject engine")
+			}
+			// find the deployment with a container for the pu marked for engine
+			for i, _ := range c.deployments {
+				dep := c.deployments[i]
+				for _, con := range dep.Spec.Template.Spec.Containers {
+					if strings.Compare(con.Name, pu.Name) == 0 {
+						deploy = dep
+						found = true
+					}
+				}
 			}
-			// Add deployment
-			c.deployments = append(c.deployments, deploy)
-		}
 
-		//Create Service for Predictor
+			if !found {
```

Comment:
No, can't see any specific way it could. For this to happen something would have had to go wrong in create the deployments - it's a fallback check. Am adding comment to make this clear.